lib/route/qdisc.c

```diff
@@ -403,6 +403,38 @@ struct rtnl_qdisc *rtnl_qdisc_get_by_parent(struct nl_cache *cache,
 	return NULL;
 }
 
+/**
+ * Search qdisc by kind
+ * @arg cache		Qdisc cache
+ * @arg ifindex		Interface index
+ * @arg kind		Qdisc kind (tbf, htb, cbq, etc)
+ *
+ * Searches a qdisc cache previously allocated with rtnl_qdisc_alloc_cache()
+ * and searches for a qdisc matching the interface index and kind.
+ *
+ * The reference counter is incremented before returning the qdisc, therefore
+ * the reference must be given back with rtnl_qdisc_put() after usage.
+ *
+ * @return pointer to qdisc inside the cache or NULL if no match was found.
+ */
+struct rtnl_qdisc *rtnl_qdisc_get_by_kind(struct nl_cache *cache,
+					    int ifindex, char *kind)
+{
+	struct rtnl_qdisc *q;
+
+	if (cache->c_ops != &rtnl_qdisc_ops)
+		return NULL;
+
+	nl_list_for_each_entry(q, &cache->c_items, ce_list) {
+		if (!strncmp(q->q_kind, kind, TCKINDSIZ) && q->q_ifindex == ifindex) {
```

Comment:
various other places already require that `q->q_kind` is a NUL terminated string. Also, it seems resonable to require the caller to pass `kind` as a NUL terminated string. Just use `strcmp()`.

Reply:
can you do the cheaper check first (for the ifindex)?

Reply:
I've reworked this function according to your comments.